File: Linked List/circular/updation/list.cpp
// Circular Singly Linked List - Updation (C++)
// Update first occurrence of a value.

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int value) : data(value), next(nullptr) {}
};

bool updateFirst(Node* head, int target, int newValue) {
    if (head == nullptr) return false;
    Node* cur = head;
    do {
        if (cur->data == target) { cur->data = newValue; return true; }
        cur = cur->next;
    } while (cur != head);
    return false;
}

void traverse(Node* head) {
    if (head == nullptr) { cout << "\n"; return; }
    Node* cur = head; do { cout << cur->data << " "; cur = cur->next; } while (cur != head); cout << "\n";
}

int main() {
    // Build: 10 -> 20 -> 30 -> back to 10
    Node* n1 = new Node(10); Node* n2 = new Node(20); Node* n3 = new Node(30);
    n1->next = n2; n2->next = n3; n3->next = n1; Node* head = n1;
    cout << "Before: "; traverse(head);
    bool ok = updateFirst(head, 20, 25);
    cout << (ok ? "Updated" : "Not found") << "\n";
    cout << "After: "; traverse(head);

    // cleanup
    Node* tail = head; if (tail != nullptr) { while (tail->next != head) tail = tail->next; tail->next = nullptr; }
    Node* cur = head; while (cur != nullptr) { Node* d = cur; cur = cur->next; delete d; }
    return 0;
}

File: Linked List/doubly/updation/list.py
# Doubly Linked List - Updation (Python)
# Update first occurrence of a value.

class Node:
    def __init__(self, value: int) -> None:
        self.data = value
        self.prev: "Node | None" = None
        self.next: "Node | None" = None


def update_first(head: "Node | None", target: int, new_value: int) -> bool:
    cur = head
    while cur is not None:
        if cur.data == target:
            cur.data = new_value
            return True
        cur = cur.next
    return False


def print_forward(head: "Node | None") -> None:
    cur = head
    while cur is not None:
        print(cur.data, end=" ")
        cur = cur.next
    print()


if __name__ == "__main__":
    # Build: 10 <-> 20 <-> 30
    n1 = Node(10); n2 = Node(20); n3 = Node(30)
    n1.next = n2; n2.prev = n1; n2.next = n3; n3.prev = n2
    head = n1

    print("Before:", end=" "); print_forward(head)
    ok = update_first(head, 20, 25)
    print("Updated" if ok else "Not found")
    print("After:", end=" "); print_forward(head)


